R/splitRaster.R

```diff
@@ -80,69 +92,139 @@ setGeneric("splitRaster", function(x, nx, ny, buffer) {
         y1 <- ext@ymin + (j + 1L)*((ext@ymax - ext@ymin) / ny) + buffer[2]*yres(x)
-        tiles[[n]] <- crop(x, extent(x0, x1, y0, y1))
+        extents[[n]] <- extent(x0, x1, y0, y1)
         n <- n + 1L
       }
     }
+    cl <- tryCatch(getCluster(), error = function(x) NULL)
+    on.exit(if (!is.null(cl)) returnCluster())
+    if (!is.null(cl)) {
+      parFun <- "clusterApply"
+      cropNew <- function(ras, x0 = x, ...) {
+        crop(x = ras,y = x0)
+      }
+      args <- list(cl = cl, x = extents,
+                   fun = cropNew, ras = x)
+
+      # args <- list(seq = extents, fun = crop, r)
+      # args1 <- list(cl = cl, y = extents, fun = crop, x = r)
+      # args <- append(list(cl = cl), args)
+    } else {
+      parFun <- "lapply"
+      args <- list(extents, FUN = crop, x = r)
+    }
+    tiles <- do.call(get(parFun), args)
+    for(i in 1:length(tiles)){
+      writeRaster(tiles[[i]], file.path(savePath, paste(names(x), "_tile", i, ".tif", sep = "")),
+                  overwrite = TRUE)
+
+      singleRaster <- raster(file.path(savePath, paste(names(x), "_tile", i, ".tif", sep = "")))
+      crs(singleRaster) <- crs(x)
+      tiles[[i]] <- singleRaster
+    }
     return(tiles)
 })
 
 #' @export
 #' @rdname splitRaster
 setMethod(
   "splitRaster",
-  signature = signature(x = "RasterLayer", nx = "numeric", ny = "numeric", buffer = "integer"),
+  signature = signature(x = "RasterLayer", nx = "numeric", ny = "numeric", buffer = "integer",
+                        savePath = "character"),
+  definition = function(x, nx, ny, buffer, savePath) {
+    return(splitRaster(x, as.integer(nx), as.integer(ny), as.numeric(buffer), savePath))
+})
+
+setMethod(
+  "splitRaster",
+  signature = signature(x = "RasterLayer", nx = "numeric", ny = "numeric", buffer = "integer",
+                        savePath = "missing"),
   definition = function(x, nx, ny, buffer) {
-    return(splitRaster(x, as.integer(nx), as.integer(ny), as.numeric(buffer)))
+    return(splitRaster(x, as.integer(nx), as.integer(ny), as.numeric(buffer), savePath = file.path(getwd, names(x))))
+  })
+
+
+#' @export
+#' @rdname splitRaster
+setMethod(
+  "splitRaster",
+  signature = signature(x = "RasterLayer", nx = "numeric", ny = "numeric", buffer = "numeric",
+                        savePath = "character"),
+  definition = function(x, nx, ny, buffer, savePath) {
+    return(splitRaster(x, as.integer(nx), as.integer(ny), buffer, savePath))
 })
 
 #' @export
 #' @rdname splitRaster
 setMethod(
   "splitRaster",
-  signature = signature(x = "RasterLayer", nx = "numeric", ny = "numeric", buffer = "numeric"),
+  signature = signature(x = "RasterLayer", nx = "numeric", ny = "numeric", buffer = "numeric",
+                        savePath = "missing"),
   definition = function(x, nx, ny, buffer) {
-    return(splitRaster(x, as.integer(nx), as.integer(ny), buffer))
+    return(splitRaster(x, as.integer(nx), as.integer(ny), buffer, savePath = file.path(getwd, names(x))))
```

Comment:
Alex,

Thanks.
I will correct these errors and create pr again.
It is interesting that the CDM check did not catch these errors.

Yong

On Thu, Jun 16, 2016 at 12:02 PM, Alex Chubaty <notifications@github.com>
wrote:

> In R/splitRaster.R
> <https://github.com/PredictiveEcology/SpaDES/pull/286#discussion_r67405058>
> :
>
> >    definition = function(x, nx, ny, buffer) {
> > -    return(splitRaster(x, as.integer(nx), as.integer(ny), buffer))
> > +    return(splitRaster(x, as.integer(nx), as.integer(ny), buffer, savePath = file.path(getwd, names(x))))
>
> should be getwd()
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/PredictiveEcology/SpaDES/pull/286/files/a733cb411ca2c41632018cd6f2372a7bd13146e0#r67405058>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AJ-1fNCYW8wIEPGnJKsWHTpvZ8a2HZ1wks5qMZ3TgaJpZM4I3FuJ>
> .
>



-- 
Yong Luo
Postdoctoral Researcher

Canadian Forest Service
Natural Resources Canada
Pacific Forestry Centre
506 Burnside Road W
Victoria, BC  V8Z 1M5
Canada
phone: +1-250-298-2566
&
Faculty of Forestry
University of British Columbia
2424 Main Mall
Vancouver, BC V6T 1Z4
Canada